landingai/vision_pipeline.py

```diff
@@ -0,0 +1,338 @@
+    _lock: Any = PrivateAttr()  # threading.Lock
+    _t: Any = PrivateAttr()  # threading.Thread
+
+    def __init__(
+        self,
+        stream_url: str,
+        motion_detection_threshold: int = 0,
+        capture_interval: Union[float, None] = None,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        stream_url : url to video source
+        motion_detection_threshold : If set to zero then motion detections is disabled. Any other value (0-100) will make the camera drop all images that don't have significant changes
+        capture_interval : If set to None, the NetworkedCamera will acquire images as fast as the source permits. Otherwise will grab new frames on the given interval
+        """
+        cap = cv2.VideoCapture(stream_url)
+        if not cap.isOpened():
+            cap.release()
+            raise Exception(f"Could not open stream ({stream_url})")
+        cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)  # Limit buffering to 1 frames
+
+        super().__init__(
+            stream_url=stream_url,
+            motion_detection_threshold=motion_detection_threshold,
+            capture_interval=capture_interval,
+        )
+        self._last_capture_time = datetime.now()
+        # FPS = 1/X
+        # self.FPS_MS = int(self.FPS * 1000)
+        self._FPS = 1 / cap.get(cv2.CAP_PROP_FPS)  # Get the source's framerate
+        self._cap = cap
+        self._lock = threading.Lock()
+        self._t = threading.Thread(target=self._reader)
+        self._t.daemon = True
+        self._t.start()
+
+    def __del__(self) -> None:
+        self._cap.release()
+
+    # grab frames as soon as they are available
+    def _reader(self) -> None:
+        while True:
+            with self._lock:
+                ret = self._cap.grab()
+                time.sleep(self._FPS)  # Limit acquisition speed
+            if not ret:
+                raise Exception(f"Connection to camera broken ({self.stream_url})")
+
+    # retrieve latest frame
+    def get_latest_frame(self) -> "FrameSet":
+        """Return the most up to date frame by dropping all by the latest frame. This function is blocking"""
+        with self._lock:
+            if self.capture_interval is not None:
+                t = datetime.now()
+                delta = (t - self._last_capture_time).total_seconds()
+                if delta <= self.capture_interval:
+                    time.sleep(delta)
+                self._last_capture_time = t
+
+            ret, frame = self._cap.retrieve()
+            if not ret:
+                raise Exception(f"Connection to camera broken ({self.stream_url})")
+            if self.motion_detection_threshold > 0:
+                if self._detect_motion(frame):
+                    return FrameSet.from_array(frame)
+                else:
+                    return FrameSet()  # Empty frame
+
+        return FrameSet.from_array(frame)
+
+    def _detect_motion(self, frame: np.ndarray) -> bool:
```

Comment:
i think this function is worth adding some tests.

Reply:
Agree. Added a TODO